libass/ass_cache.c

```diff
@@ -435,7 +446,87 @@ Cache *ass_cache_create(const CacheDesc *desc)
     return cache;
 }
 
-void *ass_cache_get(Cache *cache, void *key, void *priv)
+bool ass_cache_client_set_init(CacheClientSet *set)
+{
+    memset(set, 0, sizeof(*set));
+
+#if ENABLE_THREADS
+    if (pthread_mutex_init(&set->mutex, NULL) != 0)
+        return false;
+#endif
+
+    return true;
+}
+
+static void cache_client_done(CacheClient *client)
+{
+#if ENABLE_THREADS
+    pthread_mutex_destroy(&client->mutex);
+    pthread_cond_destroy(&client->cond);
+#endif
+
+    free(client);
+}
+
+void ass_cache_client_set_clear(CacheClientSet *set)
+{
+    CacheClient *client = set->first_client;
+
+    while (client) {
+        CacheClient *next = client->next;
+        cache_client_done(client);
```

Comment:
Can't, because they're a linked list. Deleting them individually would require making it a doubly-linked list and taking the lock; it's much simpler to just let them be deleted by the set.

Reply:
We can ensure that `first_client` is already `NULL` at this point, so deletion would be safe.